modules/apps/object/object-api/src/main/java/com/liferay/object/definition/directed/acyclic/graph/ObjectDefinitionDirectedAcyclicGraph.java

```diff
@@ -0,0 +1,881 @@
+import com.liferay.object.model.ObjectRelationship;
+import com.liferay.object.service.ObjectDefinitionLocalService;
+import com.liferay.object.service.ObjectDefinitionSettingLocalService;
+import com.liferay.object.service.ObjectEntryLocalService;
+import com.liferay.object.service.ObjectRelationshipLocalService;
+import com.liferay.object.service.persistence.ObjectActionPersistence;
+import com.liferay.object.service.persistence.ObjectDefinitionPersistence;
+import com.liferay.object.service.persistence.ObjectFieldPersistence;
+import com.liferay.object.service.persistence.ObjectRelationshipPersistence;
+import com.liferay.object.tree.Node;
+import com.liferay.object.tree.ObjectDefinitionTreeFactory;
+import com.liferay.object.tree.Tree;
+import com.liferay.petra.function.transform.TransformUtil;
+import com.liferay.petra.reflect.ReflectionUtil;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.dao.db.DB;
+import com.liferay.portal.kernel.dao.db.DBManagerUtil;
+import com.liferay.portal.kernel.dao.jdbc.CurrentConnectionUtil;
+import com.liferay.portal.kernel.dao.orm.ActionableDynamicQuery;
+import com.liferay.portal.kernel.dao.orm.RestrictionsFactoryUtil;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.exception.SystemException;
+import com.liferay.portal.kernel.model.GroupConstants;
+import com.liferay.portal.kernel.model.ResourceAction;
+import com.liferay.portal.kernel.model.ResourceConstants;
+import com.liferay.portal.kernel.model.ResourcePermission;
+import com.liferay.portal.kernel.model.WorkflowDefinitionLink;
+import com.liferay.portal.kernel.search.Indexer;
+import com.liferay.portal.kernel.search.IndexerRegistryUtil;
+import com.liferay.portal.kernel.security.permission.ResourceActions;
+import com.liferay.portal.kernel.service.ResourceActionLocalService;
+import com.liferay.portal.kernel.service.ResourcePermissionLocalService;
+import com.liferay.portal.kernel.service.WorkflowDefinitionLinkLocalService;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.ListUtil;
+
+import java.sql.Connection;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.sql.DataSource;
+
+/**
+ * @author Feliphe Marinho
+ */
+public class ObjectDefinitionDirectedAcyclicGraph {
+
+	public static ObjectDefinitionDirectedAcyclicGraph getInstance() {
+		return _objectDefinitionDirectedAcyclicGraph;
+	}
+
+	public void addEdge(
+			ObjectDefinitionLocalService objectDefinitionLocalService,
+			ObjectDefinitionPersistence objectDefinitionPersistence,
+			ObjectDefinitionSettingLocalService
+				objectDefinitionSettingLocalService,
+			ObjectEntryLocalService objectEntryLocalService,
+			ObjectRelationship objectRelationship,
+			ObjectRelationshipLocalService objectRelationshipLocalService)
+		throws PortalException {
```

Comment:
Why do we need to use LocalService and Persistence?
Could you review all the parameters from all methods please?

Reply:
I just moved the code, I noticed that core infra team use persistence over local service in some places, I'm not sure the performance benefits but I can ask to them.